dali/operators/decoder/video/video_decoder_base.h

```diff
@@ -18,118 +18,225 @@
+                             "Video decoder does not support empty input samples."));
     }
-  }
 
-
-  TensorListShape<4> ReadOutputShape() {
-    TensorListShape<4> shape(frames_decoders_.size());
-    for (size_t s = 0; s < frames_decoders_.size(); ++s) {
-      TensorShape<4> sample_shape;
-      sample_shape[0] = frames_decoders_[s]->NumFrames();
+    int batch_size = input.num_samples();
+
+    // Get frame selection parameters
+    if (start_frame_.HasValue())
+      start_frame_.Acquire(spec_, ws, batch_size);
+    if (stride_.HasValue())
+      stride_.Acquire(spec_, ws, batch_size);
+    if (sequence_length_.HasValue())
+      sequence_length_.Acquire(spec_, ws, batch_size);
+
+    frames_decoders_.resize(batch_size);
+
+    bool build_index = start_frame_.HasValue() ||
+                      stride_.HasValue() ||
+                      sequence_length_.HasValue();
+
+    // Create decoders in parallel
+    ThreadPool& thread_pool = GetThreadPool(ws);
+    for (int s = 0; s < batch_size; ++s) {
+      thread_pool.AddWork([&, s](int tid) {
+        const char* data = reinterpret_cast<const char *>(input[s].data<uint8_t>());
+        size_t size = input[s].shape().num_elements();
+        auto source_info = input.GetMeta(s).GetSourceInfo();
+        frames_decoders_[s] = CreateDecoder(data, size, build_index, source_info,
+                                            ws.has_stream() ? ws.stream() : 0);
+      });
+    }
+    thread_pool.RunAll();
+
+    // Set output shape for each sample
+    TensorListShape<4> out_shape(batch_size);
+    for (int s = 0; s < batch_size; s++) {
+      auto& decoder = frames_decoders_[s];
+      DALI_ENFORCE(decoder->IsValid(), make_string("Failed to create video decoder for \"",
+                                   decoder->Filename(), "\""));
+      auto sample_shape = out_shape.tensor_shape_span(s);
+      sample_shape[0] = sequence_length_.HasValue() ? sequence_length_[s].data[0] : frames_decoders_[s]->NumFrames();
       sample_shape[1] = frames_decoders_[s]->Height();
       sample_shape[2] = frames_decoders_[s]->Width();
       sample_shape[3] = frames_decoders_[s]->Channels();
-      shape.set_tensor_shape(s, sample_shape);
     }
-    return shape;
+
+    output_desc.resize(1);
+    output_desc[0].shape = out_shape;
+    output_desc[0].type = DALI_UINT8;
+    return true;
+  }
+
+  void RunImpl(Workspace &ws) override {
+    auto &output = ws.Output<OutBackend>(0);
+    const auto &input = ws.Input<InBackend>(0);
+    int batch_size = input.num_samples();
+    
+    // Decode samples in parallel
+    ThreadPool& thread_pool = GetThreadPool(ws);
+    for (int s = 0; s < batch_size; ++s) {
+      thread_pool.AddWork([&, s](int) {
+        int64_t num_frames = output[s].shape()[0]; // It was calculated in SetupImpl
+        int stride = stride_.HasValue() ? stride_[s].data[0] : 1;
+        int start_frame = start_frame_.HasValue() ? start_frame_[s].data[0] : 0;
```

Comment:
Why not set default values in SetupImpl?

Reply:
No need to populate another buffer.